### PR TITLE
Add new tutorial: Web Scrape Amazon with Beautiful Soup

### PR DESCRIPTION
**Pull request by ashutoshkrris**
Closes issue 

- [x] My project tutorial topic has already been approved by the Codédex team
- [x] My project tutorial file renders properly in `mdx`
- [x] The content written is my own, and not plagiarized from other sources
- [x] All code written is commented or documented thoroughly
- [x] I have already completed a self-review of the project tutorial 
- [x] I will comply with changes from the Codédex team about my project tutorial

### CODE REVIEW

File: projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx [hunk 1]
```diff
@@ -0,0 +1,200 @@
+---
+title: Web Scrape Amazon with Beautiful Soup
+author: Ashutosh Krishna
+datePublished: 2022-10-15
+description: Learn how to web scrape Amazon using Beautiful Soup in this tutorial 
+header: URL
+tags:
+  - beginner
+  - python
+  - webscraping
```
Comment (sonnynomnom) [inline projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx:10]: Great job on the metadata! This is awesome. And let's take out the `webscraping` tag.

File: projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+---
+
+# Web Scrape Amazon with Beautiful Soup
+
+<AuthorAvatar author_name="Ashutosh Krishna" author_avatar="/images/projects/authors/ashutosh_krishna.jpg" />
+
+![Header image](URL)
+
+**Prerequisites:** Python fundamentals, Basics of HTML  
+**Versions:** Python 3.10, requests 2.28.1, beautifulsoup4 4.11.1, lxml 4.9.1
```
Comment (sonnynomnom) [inline projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx:20]:
```suggestion
**Versions:** Python 3.10, requests 2.28.1, beautifulsoup4 4.11.1, lxml 4.9.1  
```

I added two spaces after so that Read Time go on a new line.

File: projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+**Read Time:** 40 minutes  
+
+## [#](#-introduction) Introduction
+
+Have you ever missed an Amazon product that was available at a discounted price? Have you ever thought of building an application that notifies you when the price of an Amazon product decreases below a certain price? If yes, let's build it together.
```
Comment (sonnynomnom) [inline projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx:25]: Like the hook!

File: projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+
+In this tutorial, we will learn to build a web scraper that scrapes Amazon product details using [Beautiful Soup](https://www.crummy.com/software/BeautifulSoup/bs4/doc/).
```
Comment (sonnynomnom) [inline projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx:27]: I think we should explain Beautiful Soup (1-2 sentences) either here or in the **What is Web Scraping?** section?

File: projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+## [#](#-what-is-web-scraping) What is Web Scraping?
+
+The process of extracting data from a website is known as web scraping. This data is collected and then exported in a way that the user will find more valuable, say an API or a CSV file. Although web scraping can be done manually, automated methods are typically preferable since they can be less expensive and perform more quickly.
+
+While web scraping isn't illegal, it can become illegal if non-publicly available data is extracted. Web scraping can sometimes be difficult as the scraper may need to go through captchas or the website keeps constantly changing.
+
+
+## [#](#-setting-up) Setting up
+
+First, open a Python code editor of your choice (VS Code recommended), and create a new Python file `scraper.py`. This is where we will be writing our code to scrape Amazon.
```
Comment (sonnynomnom) [inline projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx:42]:
```suggestion
First, open a Python code editor of your choice (VS Code recommended), and create a new Python file **scraper.py**. This is where we will be writing our code to scrape Amazon.
```

We haven't put this in the Contribution Guidelines yet, but file names on Codédex should be bold!

File: projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+    product_details = {}
+    
+    # Get the product page content and create a soup
+    page = requests.get(product_url, headers=headers)
+    soup = BeautifulSoup(page.content, features="lxml")
+``` 
+
+Now comes the most important part of the tutorial where we will extract data from the page. For the sake of this tutorial, we will extract two data - the **title** of the product and the **price** of the product.
+
+### Extracting the title of the product
```
Comment (sonnynomnom) [inline projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx:106]: Can we do title case for every header?

```suggestion
### Extracting the Title of the Product
```

File: projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+Open a product page, say [this one](https://www.amazon.com/Acer-Chromebook-Graphics-802-11ac-CB512-C1KJ/dp/B09Q9YKW15/ref=lp_16225007011_1_2) on your browser, and use the **Inspect Element** option to view the HTML content of the page. You will see the below HTML code for the title of the product.
```
Comment (sonnynomnom) [inline projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx:107]: Can we just show an image here of the Amazon Product page?

Comment (sonnynomnom) [reply projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx:107]: And then in the text, tell them how to get to the Inspect Elements. Like:

- On Windows: blah blah.
- On Mac: blah blah.

File: projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+
+    # Create an empty product details dictionary
+    product_details = {}
+    
+    # Get the product page content and create a soup
+    page = requests.get(product_url, headers=headers)
+    soup = BeautifulSoup(page.content, features="lxml")
+``` 
+
+Now comes the most important part of the tutorial where we will extract data from the page. For the sake of this tutorial, we will extract two data - the **title** of the product and the **price** of the product.
```
Comment (sonnynomnom) [inline projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx:104]:
```suggestion
Now comes the most important part of the tutorial where we will extract data from the page. For the sake of this tutorial, we will extract two data:

 - The **title** of the product.
 - The **price** of the product.
```

File: projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+price = soup.find("span", attrs={"class": "apexPriceToPay"}).get_text().strip()
+```
+
+But, you will see a problem when you print the price of the product. The extracted price will be something like **$166.00$166.00** because the parent `span` element contains two `span` elements with the price text in them. But we can clean this extracted price to get the price of the product in the following way:
+
+```python
+extracted_price = soup.find("span", attrs={"class": "apexPriceToPay"}).get_text().strip()
+price = "$" + extracted_price.split("$")[1]
+```
+We split the `extracted_price` string using the **$** symbol and select the element at index 1. We also add the dollar symbol before the price.
```
Comment (sonnynomnom) [inline projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx:136]:
```suggestion
We split the `extracted_price` string using the `$` symbol and select the element at index 1. We also add the dollar symbol before the price.
```

Also should we briefly (in a few words) explain the $ sign?

File: projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+Open a product page, say [this one](https://www.amazon.com/Acer-Chromebook-Graphics-802-11ac-CB512-C1KJ/dp/B09Q9YKW15/ref=lp_16225007011_1_2) on your browser, and use the **Inspect Element** option to view the HTML content of the page. You will see the below HTML code for the title of the product.
+
+![Extracting Title](extracting-title.png)
+
+The title is wrapped with a `span` tag with an attribute id **productTitle**. Let's use this id to extract the title of the product.
+
+```python
+title = soup.find("span", attrs={"id": "productTitle"}).get_text().strip()
+```
+We use the `find()` method to find a `span` element. We pass the id **productTitle** in a dictionary called `attrs` that accepts the attributes. The `get_text()` method returns the text in a string format. The `strip()` method is used to remove any extra whitespaces.
```
Comment (sonnynomnom) [inline projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx:116]:
```suggestion
We use the `.find()` method to find a `span` element. We pass the `productTitle` id in a dictionary called `attrs` that accepts the attributes. The `.get_text()` method returns the text in a string format. The `.strip()` method is used to remove any extra whitespaces.
```

We haven't added in the Contribution Guide, the method names have the dot in front on Codédex.

**Review by sonnynomnom — COMMENTED** [review]
Hi @ashutoshkrris! Awesome draft! Can't believe you got most of the formatting down. 🙌

Anyway, I just left my review. Let me know when you are done with the changes and another team member can take a look! 🙏

File: projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+  - webscraping
+---
+
+# Web Scrape Amazon with Beautiful Soup
+
+<AuthorAvatar author_name="Ashutosh Krishna" author_avatar="/images/projects/authors/ashutosh_krishna.jpg" />
+
+![Header image](URL)
+
+**Prerequisites:** Python fundamentals, Basics of HTML
```
Comment (Goku-kun) [inline projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx:19]: Can we keep this "Python fundamentals, HTML basics"? Just to maintain uniformity

File: projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+
+## [#](#-introduction) Introduction
+
+Have you ever missed an Amazon product that was available at a discounted price? Have you ever thought of building an application that notifies you when the price of an Amazon product decreases below a certain price? If yes, let's build it together.
+
+In this tutorial, we will learn to build a web scraper that scrapes Amazon product details using [Beautiful Soup](https://www.crummy.com/software/BeautifulSoup/bs4/doc/).
+
+The final result will look something like this:
+
+![Demo](demo.gif)
```
Comment (Goku-kun) [inline projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx:31]: Need to update this URL with the absolute URL from GitHub.

File: projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+## [#](#-scraping-amazon) Scraping Amazon
+
+Before we get started, let us import the libraries we will be using in our Python file:
+
+```python
+import requests
+from bs4 import BeautifulSoup
+```
+
+When we send an HTTP request, we send some **headers** along with the request. If you go to the Network tab of your browser and reload any web page, you can find the request headers similar to below:
```
Comment (Goku-kun) [inline projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx:63]: The second sentence can sound confusing if someone who hasn't worked with DevTools is reading this article.

We can rephrase this as: "If you open devtools in your browsers(usually accessible through Command + Option + C or Control + Shift + C or F12 key) and switch to the network tab and reload any page, you can find the request headers that are similar to the ones shown below:"

File: projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+
+![Request Headers](request-headers.png)
```
Comment (Goku-kun) [inline projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx:65]: Need to change this URL with the absolute URL.

File: projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+    pass
+```
+The function accepts a string parameter `product_url` and returns a dictionary. Let's call this dictionary `product_details`.
+
+```python
+def get_product_details(product_url: str) -> dict:
+
+    # Create an empty product details dictionary
+    product_details = {}
+```
```
Comment (Goku-kun) [inline projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx:90]: Need to link the Beautiful Soup documentation with details about the Soup object.

File: projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx (hunk 7 above)
Comment (Goku-kun) [inline projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx:107]: "Inspect Element" -> "Right click on the title of the page and select the 'Inspect Element' option from the dropdown menu"

File: projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+    page = requests.get(product_url, headers=headers)
+    soup = BeautifulSoup(page.content, features="lxml")
+``` 
+
+Now comes the most important part of the tutorial where we will extract data from the page. For the sake of this tutorial, we will extract two data - the **title** of the product and the **price** of the product.
+
+### Extracting the title of the product
+Open a product page, say [this one](https://www.amazon.com/Acer-Chromebook-Graphics-802-11ac-CB512-C1KJ/dp/B09Q9YKW15/ref=lp_16225007011_1_2) on your browser, and use the **Inspect Element** option to view the HTML content of the page. You will see the below HTML code for the title of the product.
+
+![Extracting Title](extracting-title.png)
```
Comment (Goku-kun) [inline projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx:109]: Replace the image link with the absolute URL.

File: projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx (hunk 10 above)
Comment (Goku-kun) [inline projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx:116]: The `strip()` method is used to remove any extra leading and trailing whitespaces.

File: projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx [hunk 17]
```diff
@@ -0,0 +1,200 @@
...
+The title is wrapped with a `span` tag with an attribute id **productTitle**. Let's use this id to extract the title of the product.
+
+```python
+title = soup.find("span", attrs={"id": "productTitle"}).get_text().strip()
+```
+We use the `find()` method to find a `span` element. We pass the id **productTitle** in a dictionary called `attrs` that accepts the attributes. The `get_text()` method returns the text in a string format. The `strip()` method is used to remove any extra whitespaces.
+
+### Extracting the price of the product
+
+Similar to the title of the product, if you inspect the page, you will find the below HTML code for the price of the product.
```
Comment (Goku-kun) [inline projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx:120]: Similar to the title of the product, if you inspect the prices by right clicking on them and opening devtools, you will find the below HTML code for the price of the product.

File: projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx [hunk 18]
```diff
@@ -0,0 +1,200 @@
...
+
+print(product_details)
+
+```
+
+## [#](#-conclusion) Conclusion
+
+In this tutorial, we scraped Amazon to extract the title and price of a product. You can follow the same pattern to extract more details like the rating of the product or its availability.
+
+As mentioned before, the layout of a website keeps changing and hence, the above code may not be helpful in the future. But this tutorial should be helpful to teach to you how to scrape any website.
```
Comment (Goku-kun) [inline projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx:194]: "the above code may not be helpful" -> "the above code may no longer be applicable as it is, without some changes. Join Codédex club to get support from code mentors if you run any trouble while using this article."

@sonnynomnom What do you think about this? Just the second updated sentence.

**Review by Goku-kun — CHANGES_REQUESTED** [review]
Hi, I've reviewed this article and think we need to make some more basic changes before this goes live.

File: projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx (hunk 13 above)
Comment (ashutoshkrris) [reply projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx:63]: Yes @Goku-kun 
Thanks for the suggestion.
I was going to add it after @sonnynomnom asked to.

File: projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx (hunk 18 above)
Comment (sonnynomnom) [inline projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx]: Yeah we can do that!

```suggestion
As mentioned before, the layout of Amazon will keep changing and hence, the above code may not be helpful in the future. Join Codédex Club to get support from code mentors if you run any trouble while using this article!
```

File: projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx [hunk 19]
```diff
@@ -0,0 +1,213 @@
...
+
+In this tutorial, we will learn to build a web scraper that scrapes Amazon product details using [Beautiful Soup](https://www.crummy.com/software/BeautifulSoup/bs4/doc/).
+
+The final result will look something like this:
+
+![Demo](https://raw.githubusercontent.com/codedex-io/projects/main/projects/web-scrape-amazon-with-beautiful-soup/demo.gif)
+
+## [#](#-what-is-web-scraping) What is Web Scraping?
+
+The process of extracting data from a website is known as web scraping. This data is collected and then exported in a way that the user will find more valuable, say an API or a CSV file. Although web scraping can be done manually, automated methods are typically preferable since they can be less expensive and perform more quickly.
```
Comment (asiqurrahman) [inline projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx:34]: A Hyperlink to this wiki article might be helpful for users wanting to get a deeper understanding of web scraping.

 https://en.wikipedia.org/wiki/Web_scraping

File: projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx [hunk 20]
```diff
@@ -0,0 +1,213 @@
...
+The function accepts a string parameter `product_url` and returns a dictionary. Let's call this dictionary `product_details`.
+
+```python
+def get_product_details(product_url: str) -> dict:
+
+    # Create an empty product details dictionary
+    product_details = {}
+```
+
+Now, let us make an HTTP request to get the HTML content of the product page. Using the `BeautifulSoup` class, we will create a [soup](https://www.crummy.com/software/BeautifulSoup/bs4/doc/#making-the-soup) of the page content.
```
Comment (asiqurrahman) [inline projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx:93]: For me, "let us" breaks the flow of the sentence. "Now, let's make" would sound better I think.
This is subjective of course.

File: projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx [hunk 21]
```diff
@@ -0,0 +1,213 @@
...
+Use the **Inspect** option to view the HTML code used for the title of the product.
+
+![Extracting Title](https://raw.githubusercontent.com/codedex-io/projects/main/projects/web-scrape-amazon-with-beautiful-soup/extracting-title.png)
+
+The title is wrapped with a `span` tag with an attribute id **productTitle**. Let's use this id to extract the title of the product.
+
+```python
+title = soup.find("span", attrs={"id": "productTitle"}).get_text().strip()
+```
+We use the `.find()` method to find a `span` element. We pass the `productTitle` id in a dictionary called `attrs` that accepts the attributes. The `.get_text()` method returns the text in a string format. The `.strip()` method is used to remove any extra leading and trailing whitespaces.
```
Comment (asiqurrahman) [inline projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx:129]: Replace "we" with "Then" in the second sentence.

Comment (ashutoshkrris) [reply projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx]: Done @asiqurrahman

File: projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx [hunk 22]
```diff
@@ -0,0 +1,213 @@
...
+    page = requests.get(product_url, headers=headers)
+    soup = BeautifulSoup(page.content, features="lxml")
+``` 
+
+Now comes the most important part of the tutorial where we will extract data from the page. For the sake of this tutorial, we will extract two data: 
+
+- The **title** of the product.
+- The **price** of the product.
+
+But before extracting, let us learn how to view the HTML code that makes up any web page. Open a product page, say [this one](https://www.amazon.com/Acer-Chromebook-Graphics-802-11ac-CB512-C1KJ/dp/B09Q9YKW15) on your browser.
```
Comment (asiqurrahman) [inline projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx:110]: "let us" --> "let's"

File: projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx [hunk 23]
```diff
@@ -0,0 +1,213 @@
...
+
+print(product_details)
+
+```
+
+## [#](#-conclusion) Conclusion
+
+In this tutorial, we scraped Amazon to extract the title and price of a product. You can follow the same pattern to extract more details like the rating of the product or its availability.
+
+As mentioned before, the layout of Amazon will keep changing and hence, the above code may not be helpful in the future. Join Codédex Club to get support from code mentors if you run any trouble while using this article!
```
Comment (asiqurrahman) [inline projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx:207]: "if you run any trouble" should be "if you run into any trouble".

That should be my final feedback.
Amazing article, loved the pace of it and how easy it is to understand!

Comment (ashutoshkrris) [reply projects/web-scrape-amazon-with-beautiful-soup/web-scrape-amazon-with-beautiful-soup.mdx]: Great, thank you @asiqurrahman 
I've made the required changes.

**Review by Goku-kun — APPROVED** [review]
I think we're ready to publish this article now.

I'm merging this in!

**Comment by ashutoshkrris** [comment]
Awesome
Thanks, @Goku-kun and the entire Codédex team.
